Add Mixtral and support for chat endpoints via Fireworks

philipp-spiess opened the pull request:
This PR adds support for Mixtral in both the completion mode (via an instruct tuned version) and the chat mode (via a custom chat version).

It also adds support to call the chat completions endpoints with Fireworks for the first time.

~~cc @rafax I was not able to get the API call to work with the `cody-pro` flag is on. Can you test this again and make sure it works?~~ ✅ 

~~cc @beyang @vdavid Are you fine with me adding this new open source model to the Cody Pro allowlist?~~ ✅  (I got Beyang's OK)

## Test plan

```
α dev/sourcegraph 🐙(ps/mixtral)✗
curl 'https://sourcegraph.test:3443/.api/completions/stream' \
-X POST \
-H 'authorization: token TOKEN' \
--data-raw '{"messages":[{"speaker":"human","text":"You are Cody"}, {"speaker": "assistant", "text": "Ok, I am Cody"}, {"speaker":"human","text":"Who made you though and what is your real name name though?"}],"maxTokensToSample":30,"temperature":0.2,"topP":0.95,"topK":0,"model":"fireworks/accounts/fireworks/models/mixtral-8x7b-instruct","stopSequences":[],"timeoutMs":5000,"stream":true}'
event: completion
data: {"completion":"","stopReason":""}

event: completion
data: {"completion":"I was created by Mistral","stopReason":""}

event: completion
data: {"completion":"I was created by Mistral AI, a leading AI","stopReason":""}

event: completion
data: {"completion":"I was created by Mistral AI, a leading AI company. My real name","stopReason":""}

event: completion
data: {"completion":"I was created by Mistral AI, a leading AI company. My real name is Cody, and","stopReason":""}

event: completion
data: {"completion":"I was created by Mistral AI, a leading AI company. My real name is Cody, and I am designed to assist","stopReason":""}

event: completion
data: {"completion":"I was created by Mistral AI, a leading AI company. My real name is Cody, and I am designed to assist and support you with","stopReason":"length"}

event: done
data: {}
```

rafax commented on cmd/cody-gateway/internal/httpapi/completions/fireworks.go:
```diff
@@ -70,10 +71,9 @@ func NewFireworksHandler(
-				if logSelfServeCodeCompletionRequests && act.IsDotComActor() {
+			getRequestMetadata: func(ctx context.Context, logger log.Logger, act *actor.Actor, feature codygateway.Feature, body fireworksRequest) (model string, additionalMetadata map[string]any) {
+				// Check that this is a code completion request and that the actor is a PLG user
+				if feature != codygateway.FeatureCodeCompletions && logSelfServeCodeCompletionRequests && act.IsDotComActor() {
```
```suggestion
				if feature != codygateway.FeatureCodeCompletions && logSelfServeCodeCompletionRequests && act.IsDotComActor() {
```
Should this be ==? This code logs code completions, but != sould stop it from being called

philipp-spiess replied:
YES! 😰

rafax commented on cmd/cody-gateway/internal/httpapi/completions/upstream.go:
```diff
@@ -221,7 +222,8 @@ func makeUpstreamHandler[ReqT UpstreamRequest](
 
 			// Create a new request to send upstream, making sure we retain the same context.
-			req, err := http.NewRequestWithContext(r.Context(), http.MethodPost, upstreamAPIURL, bytes.NewReader(upstreamPayload))
+			url := upstreamAPIURL(feature)
```
Nit: it seems we only use `url` once, so I'd inline it.

rafax commented:
FYI - your original issue is caused by the fact this PR https://github.com/sourcegraph/deploy-sourcegraph-cloud/commit/b7d5368710b0c6afd59f1b5c0a339e9df157ddb9 has no matching PR in dev-private (cc @thenamankumar @vdavid you need to add those new values to dev-private so local env works correctly).

philipp-spiess commented:
@rafax OOOOOOO, testing

vdavid commented:
@rafax @philipp-spiess, oh crap, I see, adding it now.

vdavid commented:
PR here: https://github.com/sourcegraph/dev-private/pull/78

philipp-spiess commented:
@rafax @vdavid I’m getting a new error here but _I think_ this is because #58932 is missing, would that make sense?


```
α dev/sourcegraph 🐙(ps/mixtral)✗
curl "http://localhost:9992/v1/limits" -H "Authorization: Bearer token"
{"chat_completions":{"limit":500,"interval":"24h0m0s","usage":0,"allowedModels":["anthropic/claude-v1","anthropic/claude-2","anthropic/claude-2.0","anthropic/claude-2.1","anthropic/claude-instant-v1","anthropic/claude-instant-1","fireworks/accounts/fireworks/models/mixtral-8x7b-fw-chat"]},"code_completions":{"limit":10000,"interval":"24h0m0s","usage":0,"allowedModels":["anthropic/claude-instant-v1","anthropic/claude-instant-1","fireworks/accounts/fireworks/models/starcoder-7b-w8a16","fireworks/accounts/sourcegraph/models/starcoder-7b","fireworks/accounts/fireworks/models/starcoder-16b-w8a16","fireworks/accounts/sourcegraph/models/starcoder-16b"]},"embeddings":{"limit":200000000,"interval":"720h0m0s","usage":0,"allowedModels":["openai/text-embedding-ada-002"]}}
```

```
α dev/sourcegraph 🐙(ps/mixtral)✗
curl 'https://sourcegraph.test:3443/.api/completions/stream' \
-X POST \
-H 'authorization: token token' \
--data-raw '{"messages":[{"speaker":"human","text":"You are Cody"}, {"speaker": "assistant", "text": "Ok, I am Cody"}, {"speaker":"human","text":"What is your real name name though?"}],"maxTokensToSample":30,"temperature":0.2,"topP":0.95,"topK":0,"model":"fireworks/accounts/fireworks/models/mixtral-8x7b-fw-chat","stopSequences":[],"timeoutMs":5000,"stream":true}'
event: error
data: {"error":"Sourcegraph Cody Gateway: unexpected status code 404: {\"error\": \"Model not found.\"}"}

event: done
data: {}

vdavid commented:
Try it with `anthropic/claude-2.0`, @philipp-spiess!

rafax commented:
> @rafax @vdavid I’m getting a new error here but _I think_ this is because #58932 is missing, would that make sense?
> 
> ```
> α dev/sourcegraph 🐙(ps/mixtral)✗
> curl "http://localhost:9992/v1/limits" -H "Authorization: Bearer token"
> {"chat_completions":{"limit":500,"interval":"24h0m0s","usage":0,"allowedModels":["anthropic/claude-v1","anthropic/claude-2","anthropic/claude-2.0","anthropic/claude-2.1","anthropic/claude-instant-v1","anthropic/claude-instant-1","fireworks/accounts/fireworks/models/mixtral-8x7b-fw-chat"]},"code_completions":{"limit":10000,"interval":"24h0m0s","usage":0,"allowedModels":["anthropic/claude-instant-v1","anthropic/claude-instant-1","fireworks/accounts/fireworks/models/starcoder-7b-w8a16","fireworks/accounts/sourcegraph/models/starcoder-7b","fireworks/accounts/fireworks/models/starcoder-16b-w8a16","fireworks/accounts/sourcegraph/models/starcoder-16b"]},"embeddings":{"limit":200000000,"interval":"720h0m0s","usage":0,"allowedModels":["openai/text-embedding-ada-002"]}}
> ```
> 
> ```
> α dev/sourcegraph 🐙(ps/mixtral)✗
> curl 'https://sourcegraph.test:3443/.api/completions/stream' \
> -X POST \
> -H 'authorization: token token' \
> --data-raw '{"messages":[{"speaker":"human","text":"You are Cody"}, {"speaker": "assistant", "text": "Ok, I am Cody"}, {"speaker":"human","text":"What is your real name name though?"}],"maxTokensToSample":30,"temperature":0.2,"topP":0.95,"topK":0,"model":"fireworks/accounts/fireworks/models/mixtral-8x7b-fw-chat","stopSequences":[],"timeoutMs":5000,"stream":true}'
> event: error
> data: {"error":"Sourcegraph Cody Gateway: unexpected status code 404: {\"error\": \"Model not found.\"}"}
> 
> event: done
> data: {}
> ```

This looks like upstream error (based on error nesting) - can you hit Fireworks directly?

rafax commented:
> > @rafax @vdavid I’m getting a new error here but _I think_ this is because #58932 is missing, would that make sense?
> > ```
> > α dev/sourcegraph 🐙(ps/mixtral)✗
> > curl "http://localhost:9992/v1/limits" -H "Authorization: Bearer token"
> > {"chat_completions":{"limit":500,"interval":"24h0m0s","usage":0,"allowedModels":["anthropic/claude-v1","anthropic/claude-2","anthropic/claude-2.0","anthropic/claude-2.1","anthropic/claude-instant-v1","anthropic/claude-instant-1","fireworks/accounts/fireworks/models/mixtral-8x7b-fw-chat"]},"code_completions":{"limit":10000,"interval":"24h0m0s","usage":0,"allowedModels":["anthropic/claude-instant-v1","anthropic/claude-instant-1","fireworks/accounts/fireworks/models/starcoder-7b-w8a16","fireworks/accounts/sourcegraph/models/starcoder-7b","fireworks/accounts/fireworks/models/starcoder-16b-w8a16","fireworks/accounts/sourcegraph/models/starcoder-16b"]},"embeddings":{"limit":200000000,"interval":"720h0m0s","usage":0,"allowedModels":["openai/text-embedding-ada-002"]}}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > α dev/sourcegraph 🐙(ps/mixtral)✗
> > curl 'https://sourcegraph.test:3443/.api/completions/stream' \
> > -X POST \
> > -H 'authorization: token token' \
> > --data-raw '{"messages":[{"speaker":"human","text":"You are Cody"}, {"speaker": "assistant", "text": "Ok, I am Cody"}, {"speaker":"human","text":"What is your real name name though?"}],"maxTokensToSample":30,"temperature":0.2,"topP":0.95,"topK":0,"model":"fireworks/accounts/fireworks/models/mixtral-8x7b-fw-chat","stopSequences":[],"timeoutMs":5000,"stream":true}'
> > event: error
> > data: {"error":"Sourcegraph Cody Gateway: unexpected status code 404: {\"error\": \"Model not found.\"}"}
> > 
> > event: done
> > data: {}
> > ```
> 
> This looks like upstream error (based on error nesting) - can you hit Fireworks directly?

curl 'https://api.fireworks.ai/inference/v1/chat/completions' \
  -H 'authority: api.fireworks.ai' \
  -H 'accept: text/event-stream' \
  -H 'accept-language: en-GB,en-US;q=0.9,en;q=0.8' \
  -H 'authorization: Bearer TOKEN' \
  -H 'content-type: application/json' \
  -H 'origin: https://app.fireworks.ai' \
  -H 'sec-ch-ua: "Google Chrome";v="117", "Not;A=Brand";v="8", "Chromium";v="117"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "macOS"' \
  -H 'sec-fetch-dest: empty' \
  -H 'sec-fetch-mode: cors' \
  -H 'sec-fetch-site: same-site' \
  -H 'user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Safari/537.36' \
  --data-raw $'{"model":"accounts/fireworks/models/mixtral-8x7b-fw-chat","messages":[{"role":"user","content":"Hey\u0021\\n"}],"stream":true,"n":1,"max_tokens":150,"temperature":0.1,"top_p":0.9,"logprobs":1}' \
  --compressed
{"error": "Model not found."}%

philipp-spiess commented:
Moved to the `-instruct` model now and updated the test plan. Waiting from a quick OK from Fireworks too (https://sourcegraph.slack.com/archives/C0514L18QHG/p1702478753388689) but got an OK to add this as an allowed cody pro model for GA from Beyang. Do you have any concerns for adding this @vdavid?

rafax commented:
I think you need this to make sure people that get a new client _before_ we launch the Cody Pro flag for everyone can use it 
```
diff --git a/cmd/frontend/internal/dotcom/productsubscription/codygateway_dotcom_user.go b/cmd/frontend/internal/dotcom/productsubscription/codygateway_dotcom_user.go
index e72f79939a..385a3e0d35 100644
--- a/cmd/frontend/internal/dotcom/productsubscription/codygateway_dotcom_user.go
+++ b/cmd/frontend/internal/dotcom/productsubscription/codygateway_dotcom_user.go
@@ -339,7 +339,7 @@ func allowedModels(scope types.CompletionsFeature, isCodyProEnabled, isProUser b
        switch scope {
        case types.CompletionsFeatureChat:
                if !isCodyProEnabled {
-                       return []string{"anthropic/claude-v1", "anthropic/claude-2", "anthropic/claude-2.0", "anthropic/claude-2.1", "anthropic/claude-instant-v1", "anthropic/claude-instant-1"}
+                       return []string{"anthropic/claude-v1", "anthropic/claude-2", "anthropic/claude-2.0", "anthropic/claude-2.1", "anthropic/claude-instant-v1", "anthropic/claude-instant-1", "fireworks/accounts/fireworks/models/mixtral-8x7b-instruct"}
                }

                // When updating the below lists, make sure you also update `isAllowedCustomChatModel` in `chat.go`
```

rafax commented on cmd/frontend/internal/dotcom/productsubscription/codygateway_dotcom_user.go:
Also add on line 342

rafax reviewed (approved):
Code LGTM, please:

- update the sourcegraph.com subscription
- let me know how you plan to roll this out and test live (will you do a one-off deploy of sourcegraph.com? wait for tomorrow?)

philipp-spiess commented:
@rafax 

> update the sourcegraph.com subscription

done

> let me know how you plan to roll this out and test live (will you do a one-off deploy of sourcegraph.com? wait for tomorrow?)

I was thinking of doing an ad-hoc rollout. I can take care of it. Just want to get as much testing time as possible since we plan to ship tomorrow
